File: frontend/src/api/content.ts
import { apiClient } from './client';
import { Generation } from '../types';

type BackendScene = {
  scene: string;
  time: string;
};

type BackendGeneration = {
  generation_id: number;
  product_id: number;
  viral_score: number;
  trend?: Record<string, unknown> | null;
  hook: string;
  meme_format: string;
  dialogue: string[];
  script: BackendScene[];
  caption: string;
  hashtags: string[];
  klipy_query: string;
  meme?: {
    title?: string | null;
    url?: string | null;
    preview?: string | null;
    source?: string | null;
    mediaType?: string | null;
    providerStatus?: string | null;
  } | null;
};

export const normalizeGeneration = (data: BackendGeneration): Generation => {
  const trendName = String(data.trend?.name || data.trend?.title || data.klipy_query || 'Generated trend');
  const trendReason = String(data.trend?.reason || data.trend?.analysis || 'Selected from the generated brief and current content signals.');
  const mediaUrl = data.meme?.url || data.meme?.preview || '';

  return {
    generation_id: data.generation_id,
    viral_score: data.viral_score,
    trend: {
      format: trendName,
      reason: trendReason,
    },
    hook: data.hook,
    meme: {
      format: data.meme_format,
      search_query: data.klipy_query,
      gif: mediaUrl
        ? {
            id: data.meme?.title || data.klipy_query,
            title: data.meme?.title || data.meme_format,
            preview_url: data.meme?.preview || mediaUrl,
            gif_url: mediaUrl,
          }
        : null,
    },
    dialogue: (data.dialogue || []).map((line, index) => ({
      speaker: index % 2 === 0 ? 'Person A' : 'Person B',
      line,
    })),
    script: (data.script || []).map((scene) => ({
      time: scene.time,
      visual: scene.scene,
      voice: scene.scene,
      text_overlay: data.hook,
    })),
    caption: data.caption,
    hashtags: data.hashtags || [],
  };
};

export const generateContent = async (productId: number, tone?: string, requirements?: string): Promise<Generation> => {
  const response = await apiClient.post('/api/content/generate', {
    product_id: productId,
    tone,
    requirements,
  });
  return normalizeGeneration(response.data);
};

File: backend/schemas/content.py
"""
schemas/content.py — AI output schema and content generation request/response.
"""
from typing import List, Optional

from pydantic import BaseModel, Field


# ── AI structured output schema (sent to Gemini as response_schema) ──────────

class TrendInfo(BaseModel):
    format: str = Field(..., description="Trend format name, e.g. 'POV'")
    reason: str = Field(..., description="Why this trend fits the product")


class MemeInfo(BaseModel):
    format: str = Field(..., description="Meme format name")
    search_query: str = Field(..., description="Search query for Tenor GIF")


class DialogueLine(BaseModel):
    speaker: str = Field(..., description="Who is speaking")
    line: str = Field(..., description="Dialogue line")


class ScriptScene(BaseModel):
    time: str = Field(..., description="Timestamp range, e.g. '0-3 sec'")
    visual: str = Field(..., description="What the viewer sees")
    voice: str = Field(..., description="Voiceover or spoken text")
    text_overlay: str = Field(..., description="On-screen text overlay")


class AIResponse(BaseModel):
    """Exact schema Gemini must conform to."""
    viral_score: int = Field(..., ge=0, le=100, description="Predicted virality 0-100")
    trend: TrendInfo
    hook: str = Field(..., description="Attention-grabbing opening line")
    meme: MemeInfo
    dialogue: List[DialogueLine] = Field(default_factory=list)
    script: List[ScriptScene] = Field(default_factory=list)
    caption: str = Field(..., description="Social media post caption")
    hashtags: List[str] = Field(default_factory=list, description="Relevant hashtags")


# ── Tenor GIF result ──────────────────────────────────────────────────────────

class GifResult(BaseModel):
    id: str
    title: str
    preview_url: str
    gif_url: str
    width: Optional[int] = None
    height: Optional[int] = None


# ── Meme section in the full response (adds optional GIF) ────────────────────

class MemeWithGif(BaseModel):
    format: str
    search_query: str
    gif: Optional[GifResult] = None


# ── Generation request ────────────────────────────────────────────────────────

class GenerateRequest(BaseModel):
    product_id: int = Field(..., description="ID of the product to generate content for")

    model_config = {"json_schema_extra": {"example": {"product_id": 1}}}


# ── Full content generation response ─────────────────────────────────────────

class ContentResponse(BaseModel):
    generation_id: int
    viral_score: int
    trend: TrendInfo
    hook: str
    meme: MemeWithGif
    dialogue: List[DialogueLine]
    script: List[ScriptScene]
    caption: str
    hashtags: List[str]
